File: AlgorithmDataStructure/LeetCode/2-AddTwoNumber/add.c

#include <stdlib.h>
#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};


struct ListNode *listCreate(int val) {
    struct ListNode *list = NULL;
    list = malloc(sizeof(*list));
    if (list == NULL) {
        return NULL;
    }
    list->val = val;
    list->next = NULL;
    return list;
}

void addNodeToTail(struct ListNode *list, int val) {
    if (list == NULL) {
        return;
    }
    
    struct ListNode *node = malloc(sizeof(struct ListNode));
    if (node == NULL) {
        return;
    }
    node->val = val;
    node->next = NULL;
    
    struct ListNode *tmpNode = list;
    while (tmpNode->next != NULL) {
        tmpNode = tmpNode->next;
    }
    tmpNode->next = node;
}

void listDump(struct ListNode *list) {
    if (list == NULL) {
        printf("empty list!\n");
    }
    struct ListNode *tmpNode = list;
    while (tmpNode !=  NULL) {
        printf("%d -> ", tmpNode->val);
        tmpNode = tmpNode->next;
    }
    printf("\n--------\n");
}


struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    
    struct ListNode *result = malloc(sizeof(struct ListNode));
    if (result == NULL) {
        return NULL;
    }
    struct ListNode *tempResult = result;
    int flag = 0, sum = 0;
    
    while (l1 || l2) {
        struct ListNode *node = malloc(sizeof(struct ListNode));
        node -> next = NULL;
        
        sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + flag;
        if (sum >= 10) {
            sum = sum%10;
            flag = 1;
        } else {
            flag = 0;
        }
        node -> val = sum;
        
        tempResult -> next = node;
        tempResult = tempResult -> next;
        
        l1 = l1 ? l1->next : NULL;
        l2 = l2 ? l2->next : NULL;
    }
    
    if (flag == 1) {
        struct ListNode *node = malloc(sizeof(struct ListNode));
        node -> next = NULL;
        node -> val = 1;
        tempResult -> next = node;
    }
    
    return result->next;
}



int main(int argc, const char * argv[]) {
    
    struct ListNode *l1 = listCreate(1);
//    addNodeToTail(l1, 4);
//    addNodeToTail(l1, 3);
//    listDump(l1);

    struct ListNode *l2 = listCreate(9);
    addNodeToTail(l2, 9);
//    addNodeToTail(l2, 4);
//    listDump(l2);
    
    struct ListNode *result = addTwoNumbers(l1, l2);
    listDump(result);
    //0 -> 0 -> 1
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/2-AddTwoNumber/add.swift

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var result: ListNode = ListNode(0)
        let tmpResult = result
        var flag = 0, sum = 0, tmpl1 = l1, tmpl2 = l2;
        while (tmpl1 != nil) || (tmpl2 != nil) {
            let node = ListNode(0)
            sum = (tmpl1?.val ?? 0) + (tmpl2?.val ?? 0) + flag
            if (sum > 9) {
                sum = sum%10
                flag = 1
            } else {
                flag = 0
            }
            node.val = sum
            node.next = nil
            
            result.next = node
            result = (result.next)!
            
            tmpl1 = tmpl1?.next
            tmpl2 = tmpl2?.next
        }
        
        if flag == 1 {
            result.next = ListNode(1)
        }
        
        return tmpResult.next
    }
}

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

let s = Solution()
var a1 = ListNode(5)
var a2 = ListNode(5)
var r = s.addTwoNumbers(a1, a2)
print("\(r?.val ?? -1) \(r?.next?.val ?? -1) ") // 0 1
